codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -89,3 +94,149 @@ let fetchNext = async (
+let waitForNewBlock = async (sourceManager: t, ~currentBlockHeight) => {
+  let {stalledPollingInterval, sources} = sourceManager
+
+  let logger = Logging.createChild(
+    ~params={
+      "chainId": sourceManager.activeSource.chain->ChainMap.Chain.toChainId,
+      "currentBlockHeight": currentBlockHeight,
+    },
+  )
+  logger->Logging.childTrace("Initiating check for new blocks.")
+
+  let syncSources = []
+  let fallbackSources = []
+  sources->Array.forEach(source => {
+    if (
+      source.sourceFor === Sync ||
+        // Even if the active source is a fallback, still include
+        // it to the list. So we don't wait for a timeout again
+        // if all main sync sources are still not valid
+        source === sourceManager.activeSource
+    ) {
+      syncSources->Array.push(source)
+    } else {
+      fallbackSources->Array.push(source)
+    }
+  })
+
+  let status = ref(Active)
+
+  let (source, newBlockHeight) = await Promise.race(
+    syncSources
+    ->Array.map(async source => {
+      (
+        source,
+        await getSourceNewHeight(
+          ~source,
+          ~currentBlockHeight,
+          ~status,
+          ~logger,
+          ~stalledPollingInterval,
+        ),
+      )
+    })
+    ->Array.concat([
+      Utils.delay(sourceManager.newBlockFallbackStallTimeout)->Promise.then(() => {
+        if status.contents !== Done {
+          status := Stalled
+
+          switch fallbackSources {
+          | [] =>
+            logger->Logging.childError(
+              `No new blocks detected within ${(sourceManager.newBlockFallbackStallTimeout / 1000)
+                  ->Int.toString}s. Polling will continue at a reduced rate. For better reliability, refer to our RPC fallback guide: https://docs.envio.dev/docs/HyperIndex/rpc-sync`,
+            )
+          | _ =>
+            logger->Logging.childWarn(
+              `No new blocks detected within ${(sourceManager.newBlockFallbackStallTimeout / 1000)
+                  ->Int.toString}s. Continuing polling with fallback RPC sources from the configuration.`,
+            )
+          }
+        }
+        // Promise.race will be forever pending if fallbackSources is empty
+        // which is good for this use case
+        Promise.race(
+          fallbackSources->Array.map(async source => {
+            (
+              source,
+              await getSourceNewHeight(
+                ~source,
+                ~currentBlockHeight,
+                ~status,
+                ~logger,
+                ~stalledPollingInterval,
+              ),
+            )
+          }),
+        )
+      }),
+    ]),
+  )
+
+  sourceManager.activeSource = source
+
+  // Show a higher level log if we displayed a warning/error after newBlockFallbackStallTimeout
+  let log = status.contents === Stalled ? Logging.childInfo : Logging.childTrace
+  logger->log({
+    "msg": `New blocks successfully found.`,
+    "source": source.name,
+    "newBlockHeight": newBlockHeight,
+  })
```

Comment:
This is what we discussed for other places to show a success log after a warning.